### PR TITLE
[feature] Allow math operations in offset modifier

### LINKED ISSUE #228: Allow basic math operations in offset modifier
The offset modifier should allow to define an offset with this fingerprint:

```
'10px + 15vh - 10 20% + 4vw - 4px'
```

I'm already working on it.

### PR DESCRIPTION
**Pull request by FezVrasta**
fix #228, pinging @shmax for review.

New documentation:

> The `offset` modifier can shift your popper on both its axis.
> 
 > It accepts the following units:
>  - `px` or unitless, interpreted as pixels
>  - `%` or `%r`, percentage relative to the length of the reference element
>  - `%p`, percentage relative to the length of the popper element
>  - `vw`, CSS viewport width unit
> - `vh`, CSS viewport height unit
> 
>  For length is intended the main axis relative to the placement of the popper.
>  This means that if the placement is `top` or `bottom`, the length will be the
>  `width`. In case of `left` or `right`, it will be the height.
> 
>  You can provide a single value (as `Number` or `String`), or a pair of values
>  as `String` divided by a comma or one (or more) white spaces.
> The latter is a deprecated method because it leads to confusion and will be
removed in v2.
>  Additionally, it accepts additions and subtractions between dfferent unit.
>  Note that multiplications and divisions aren't spported.
> 
>  Valid examples are:
>  ```
>  10
>  '10%'
>  '10, 10'
>  '10%, 10'
>  '10 + 10%'
>  '10 - 5vh + 3%'
>  '-10px + 5vh, 5px - 6%'
>  ```

### CODE REVIEW

**Comment by shmax** [comment]
I'll have to look at this in more detail after work, but my first impression is that the syntax for the compound units is a bit loose and your tests for this feature could be a bit more comprehensive. 

For example, you have this in your tests: `5 - 25% 10px + 25%`. Visually, it's a bit difficult to parse; there are 5 spaces in it, but the third space is meant to be a divider between the axes. It's not too difficult to imagine it getting even more elaborate:

`5 - 25% + 5 -10px - 10 + 25%`

If you're absolutely sure that your regex can handle it then okay, but I'm wondering if it might be helpful to make the syntax a little tighter. Some examples:

`5-25%+5 -10px+25% // allow only one space` 

`5 - 25% + 5, -10px + 25%` // require a comma

As for your tests, I would probably think about adding some unit tests (not functional) and really beating on your parsing function. Feed it all kinds of weird scenarios.

And finally a comment on the tests you added in this PR: they seem to have a little too much logic going on. Your `expectedPopperLeft` value, for instance, is constructed from a fairly elaborate mathematical expression, which means it could have bugs. You should strive to remove any trace of logic from your tests, make them more comprehensive, and easier to read.

Need to get to work, but this is a great start!

**Comment by FezVrasta** [comment]
About the regex, I'm quite confident it can handle the cases you mentioned (I tried it). I would like to make it simpler using an array or an object with `x` and `y` (or a comma separated as you said). But it would break the current API and I don't want to release a new major version for this change.

About adding unit tests you are right, it would make it easier to test the function, I'll add them.

**Comment by FezVrasta** [comment]
@shmax I added some unit tests and I also deprecated the space separated offset in favor of a comma separated string.

If you use the space as separator it will throw a warning but will still work.

File: tests/unit/parseOffset.js [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+
+  it('parses `10px + 5%, 10px - 5%`', () => {
+    expect(
+      parseOffset(
+        '10px + 5%, 10px - 5%',
+        referenceOffsets,
+        popperOffsets,
+        'bottom'
+      )
+    ).toEqual([10 + 5, 10 - 5]);
```
Comment (shmax) [inline tests/unit/parseOffset.js]: You lost me on this one; `10 + 5` is only the correct answer if `5%` of some value is `5`, right?

Comment (shmax) [inline tests/unit/parseOffset.js]: Ah, I see it is, okay

Comment (shmax) [inline tests/unit/parseOffset.js]: Also, I know you're deprecating it, but I don't see much in the way of tests for comma-less syntax...

Comment (FezVrasta) [inline tests/unit/parseOffset.js]: I don't wanted to add too much tests for a deprecated syntax, feel free to send an additional PR to add more tests if you prefer.

Comment (shmax) [inline tests/unit/parseOffset.js]: Also, what about weird cases like:

```
100%p - 10%r, 100% // mixing of p and r
100% + 50% + 10%, 100% // stringing more than one add
100%+50px, 100px-10+10% // tight syntax
```

Comment (FezVrasta) [inline tests/unit/parseOffset.js]: Yes that's something we should add, thank you for the comment.

Comment (FezVrasta) [inline tests/unit/parseOffset.js]: The tests you suggested caught a new bug 🙃

Now I think we are well covered.

**Comment by shmax** [comment]
Yee-haw, nice! LGTM

**Comment by mqklin** [comment]
@FezVrasta how can I offset relative to cross-axis?

**Comment by mqklin** [comment]
@FezVrasta I mean this: 
`For length is intended the main axis relative to the placement of the popper.
This means that if the placement is top or bottom, the length will be the width. In case of left or right, it will be the height.`
How can I use `cross axis`, not `main axis` for this purpose?

**Comment by FezVrasta** [comment]
I may be wrong but I don't think this is implemented 🤔

**Comment by mqklin** [comment]
Sad! Anyway, I found a workaround - just use get `getBoundingClientRect` from `reference` element and put `top: ${rect.height / 2}px;` to tooltip's css.

**Comment by FezVrasta** [comment]
I think we could directly get rid of all this stuff and make `offset` accept a function that provides the `rect` of `popper` and `reference` as arguments... That would be easier and more powerful.

```js
{
  offset: (reference, popper) => [reference.width / 2, popper.height * reference.height]
}
```

**Comment by mqklin** [comment]
Brilliant idea!

**Pull request merged**